src/software/new_geom/util/almost_equal.cpp

```diff
@@ -0,0 +1,52 @@
+#include "software/new_geom/util/almost_equal.h"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <cstring>
+
+#include "software/new_geom/geom_constants.h"
+
+int64_t ulpsDistance(double a, double b);
+
+bool almostEqual(double a, double b, double fixedEpsilon, int ulpsEpsilon)
+{
+    // Handle the near-zero case
+    const double difference = fabs(a - b);
+    if (difference <= fixedEpsilon)
+        return true;
+
+    return ulpsDistance(a, b) <= ulpsEpsilon;
+}
+
+int64_t ulpsDistance(double a, double b)
```

Comment:
super duper nit: Why not define this function at the top of the file?

Reply:
Fixed